Core/GDCore/IDE/Events/EventsFunctionSelfCallChecker.cpp

```diff
@@ -0,0 +1,175 @@
+#include "GDCore/Events/Parsers/ExpressionParser2NodeWorker.h"
+#include "GDCore/Extensions/Metadata/ExpressionMetadata.h"
+#include "GDCore/Extensions/Metadata/MetadataProvider.h"
+#include "GDCore/Extensions/Metadata/ParameterMetadataTools.h"
+#include "GDCore/Extensions/PlatformExtension.h"
+#include "GDCore/IDE/Events/ArbitraryEventsWorker.h"
+#include "GDCore/Project/EventsBasedBehavior.h"
+#include "GDCore/Project/EventsBasedObject.h"
+#include "GDCore/Project/EventsFunction.h"
+#include "GDCore/Project/EventsFunctionsExtension.h"
+#include "GDCore/Project/Object.h"
+#include "GDCore/Project/Project.h"
+#include "GDCore/String.h"
+#include "GDCore/Tools/Log.h"
+
+namespace gd {
+
+/**
+ * \brief Check the type of the first condition or action.
+ */
+class GD_CORE_API FirstInstructionTypeChecker
+    : public ReadOnlyArbitraryEventsWorker {
+public:
+  FirstInstructionTypeChecker(const gd::String &eventFunctionType_,
+                              const bool isEventFunctionCondition_)
+      : eventFunctionType(eventFunctionType_),
+        isEventFunctionCondition(isEventFunctionCondition_),
+        isOnlyCallingItself(false){};
+  virtual ~FirstInstructionTypeChecker(){};
+
+  void DoVisitInstruction(const gd::Instruction &instruction,
+                          bool isCondition) override {
+    if (isCondition == isEventFunctionCondition) {
+      isOnlyCallingItself = instruction.GetType() == eventFunctionType;
+      StopAnyEventIteration();
+    }
+  };
+
+  bool isOnlyCallingItself;
+
+private:
+  gd::String eventFunctionType;
+  bool isEventFunctionCondition;
+};
+
+/**
+ * \brief Check the type of the first condition or action.
+ */
+class GD_CORE_API FirstActionExpressionTypeChecker
+    : public ReadOnlyArbitraryEventsWorker,
+      ExpressionParser2NodeWorker {
+public:
+  FirstActionExpressionTypeChecker(const gd::Platform &platform_,
+                                   const gd::String &eventFunctionType_)
+      : platform(platform_), eventFunctionType(eventFunctionType_),
+        isOnlyCallingItself(false){};
+  virtual ~FirstActionExpressionTypeChecker(){};
+
+  void DoVisitInstruction(const gd::Instruction &instruction,
+                          bool isCondition) override {
+    // Typically, it should be a "return" action.
+    if (!isCondition) {
+      gd::String lastObjectParameter = "";
+      const gd::InstructionMetadata &instrInfos =
+          MetadataProvider::GetActionMetadata(platform, instruction.GetType());
+      for (std::size_t pNb = 0; pNb < instrInfos.parameters.size(); ++pNb) {
+
+        if (ParameterMetadata::IsExpression(
+                "number", instrInfos.parameters[pNb].GetType()) ||
+            ParameterMetadata::IsExpression(
+                "string", instrInfos.parameters[pNb].GetType())) {
+          auto node = instruction.GetParameter(pNb).GetRootNode();
+          node->Visit(*this);
+        }
+      }
+      StopAnyEventIteration();
```

Comment:
Why this?
Is this to detect just the case where there is only the first action calling itself?

Reply:
Yes, it makes the check O(event count). This way it can be done at every rendering to avoid a state + signals.